python/ray/data/dataset.py

```diff
@@ -827,14 +828,115 @@ def groupby(self, key: "GroupKeyT") -> "GroupedDataset[T]":
         from ray.data.grouped_dataset import GroupedDataset
         return GroupedDataset(self, key)
 
-    def aggregate(self, *aggs: AggregateFn) -> U:
+    def agg(self, aggs: Union[Union[str, AggregateFn], List[Union[
+            str, AggregateFn]], Dict[str, Union[str, AggregateFn]]]):
+        """Aggregate using one or more operations over the specified columns.
+
+        Examples:
+            >>> ray.data.range(100).agg("sum")
+            >>> ray.data.range(100).agg(["sum", Std(ddof=0)])
+            >>> ray.data.from_items([
+                        {"A": x, "B": x**2} for x in range(100)]) \
+                    .agg({
+                        "A": ["mean", Std(ddof=0)],
```

Comment:
I'm not sure about this. I feel like we don't need to change the existing AggregateFn classes like Mean, Std, etc. When people use them directly, they need to provide the `on`. Basically AggregateFn should be self contained. What we are doing is just adding syntax sugar on top of it. Should it just be 
```
def agg(self, aggs: Union[str, Dict[str, List[str]], List[str], AggregateFn, List[AggregateFn]):
```
`str, Dict[str, List[str]], List[str]` are just syntax sugar that will be boiled down to `List[AggregateFn]`. Anything cannot be done through the syntax sugar, people can just use the advanced `List[AggregateFn]` including their own AggregateFn implementations. What do you think?

Reply:
I think that where we're not aligning here is that I'm taking a common-case API-centric view while I think that you're taking an `AggregateFn`-centric view. Users will mostly use this Pandas-like multi-aggregation API or the aggregation-specific APIs (`.mean()`), and will mostly not have to think about `AggregateFn`. The only time that they should have to use `AggregateFn` is if they're trying to do a multi-aggregation with a custom configuration of the aggregation (e.g. `Std(ddof=0)`), and for that case, I think that:
```python
ds.agg({
    "A": ["mean", Std(ddof=0), "min", "max"],
    "B": ["sum", "max"],
})
```
is much better UX than
```python
ds.agg([Mean("A"), Std("A", ddof=0), Min("A"), Max("A"), Sum("B"), Max("B")])
```

I'm also wanting to get the multi-aggregation API as close to the Pandas API as possible, which allows for aggregation functions in the column --> aggregations dict API.

>When people use them directly, they need to provide the on. Basically AggregateFn should be self contained.

I agree that the current use of `AggregateFn` is messy, but I think that's more of an issue with the `AggregateFn` abstraction design rather than its use. As I alluded to [here](https://github.com/ray-project/ray/pull/20090#discussion_r746149156), the lifetime of `on` doesn't line up with the rest of the `AggregateFn`, and is also conceptually distinct from the rest of `AggregateFn`, so it would be great if we could decouple these. Conceptually, I'd think of it as:

1. `AggregateFn` type: An implementation of the accumulator-based aggregation API.
2. `AggregateFn` instance: A configuration of an aggregation function.
3. `AggregateFn` instance + `on`: Specifies the data subset (e.g. columns) to apply the aggregation to.

`on` could technically be applied to the data before it's even given to the aggregation function (`agg_fn.accumulate()` in particular). Here's an abstraction design that I think might be better:
```python
class Aggregation:
    """A named aggregation on some subset of a dataset.
    """
    agg_fn: AggregateFn
    on_fn: Callable
    name: str

    def __init__(self, agg_fn: AggregateFn, on: Union[str, Callable], name: Optional[str] = None):
        self.agg_fn = agg_fn
        self.on_fn = _to_on_fn(on)
        if name is None:
            name = type(agg_fn).__name__.lower()
            if isinstance(on, str):
                name += "_{on}"
            else:
                name += "_{on.__name__}"
        self.name = name

class AggregateFn:
    init: Callable[[KeyType], AggType]
    accumulate: Callable[[AggType, T], AggType]
    merge: Callable[[AggType, AggType], AggType]
    finalize: Callable[[AggType], U] = lambda a: a

# Usage in `Block.combine()`

accumulators = [agg.agg_fn.init(next_key) for agg in aggs]
for r in gen():
    for i in range(len(aggs)):
        accumulators[i] = aggs[i].agg_fn.accumulate(
            accumulators[i], aggs[i].on_fn(r))
```

Decoupling the `name` and `AggregateFn` will also open the door for easily supporting [named aggregation](https://pandas.pydata.org/docs/user_guide/groupby.html#groupby-aggregate-named), which is a nice feature.

Reply:
ok, I think I understand our key difference:
In my view, AggregateFn is the core of aggregation where the aggregation logic is defined by `init, accumulate, merge, finalize`, there is no concept of `on` here since it's embedded inside `accumulate` lambda (like accumulate takes a row and it can do whatever to extract the on). So to me, `on` and predefined max, mean, etc are all syntax sugar that help users to create a list of AggregateFn.

In your view, `on` is also the core of aggregation. AggregateFn is just the algorithm and we can combine it with different `on` to create different `Aggregation` to do.

I think your way of abstraction is better: basically we decouple aggregation algorithm and the actual data we aggregate on. So AggregateFn is just the pure algorithm and Aggregate contains information about a particular instance of aggregation like `on` and `name`.

Reply:
>In your view, on is also the core of aggregation. AggregateFn is just the algorithm and we can combine it with different on to create different Aggregation to do.

Yeah that's what I'm thinking!

>I think your way of abstraction is better: basically we decouple aggregation algorithm and the actual data we aggregate on. So AggregateFn is just the pure algorithm and Aggregate contains information about a particular instance of aggregation like on and name.

I'll try this out real quick.

Reply:
@jjyao Implemented in this commit: https://github.com/ray-project/ray/pull/20090/commits/79ca0facabae235311139db70ccbf1f5bdc2962d